### PR TITLE
modern bootstrap

### PR DESCRIPTION
**Pull request by dg**
Encapsulation of bootstrap.php into class Bootstrap.

Benefits:

- code completition:

image

- removed unusual `$container = require ...`

- Possibility to setup environment without creating container `Bootstrap::boot()`

- Possibility to parameterize bootstrap `Bootstrap::boot('project-name')` or container `Bootstrap::createContainer('project-name')` 

- etc…

### CODE REVIEW

**Comment by SendiMyrkr** [comment]
I don't see any advantage when you use this static-way. You just take code and put it to another file with no benefits.

**Comment by SendiMyrkr** [comment]
And other risk is that someone take it bad way a wil be seduced to use it as global access to container

**Comment by dg** [comment]
> And other risk is that someone take it bad way a wil be seduced to use it as global access to container

Whole Nette is one big risk that someone uses it bad way. But what you write cannot be done.

**Comment by f3l1x** [comment]
**API**

```
return ConfiguratorBuilder::create()
	->setJsonError500()
	->addParameters(['scope' => 'api'])
	->addEnvironmentConfig()
	->addConfig(__DIR__ . '/config/entrypoint/api.neon')
	->addConfig(__DIR__ . '/config/config.local.neon')
	->addCustomConfig()
	->createContainer();
```

**web**

```
return ConfiguratorBuilder::create()
	->setWebError500()
	->addParameters(['scope' => 'web'])
	->addEnvironmentConfig()
	->addConfig(__DIR__ . '/config/entrypoint/web.neon')
	->addConfig(__DIR__ . '/config/config.local.neon')
	->addCustomConfig()
	->createContainer();
```

**Comment by bckp** [comment]
Same here, we have class Bootstrap, that do exactly same thing... so 👍 for me

**Comment by milo** [comment]
@dg To better imagine new concept... how would look like following code?
```php
require __DIR__ . '/../vendor/autoload.php';

$configurator = new Nette\Configurator;

# REST API
if (strpos($_SERVER['HTTP_ACCEPT'] ?? '', 'application/json') !== false) {
	$configurator->setDebugMode(false);

# CLI
} elseif (PHP_SAPI === 'cli') {
	$configurator->setDebugMode(true);

# Web
} else {
	$configurator->setDebugMode([
		'192.168.0.2', '192.168.1.2', 
	]);
}
$configurator->enableTracy(__DIR__ . '/../log', 'miloslav.hula@...');

$configurator->setTimeZone('Europe/Prague');
$configurator->setTempDirectory(__DIR__ . '/../temp');

$configurator->createRobotLoader()
	->addDirectory(__DIR__)
	->addDirectory(__DIR__ . '/../lib')
	->register();

array_push(Tracy\Debugger::getBlueScreen()->keysToHide, 'clientSecret', 'secretKey', 'signatureKey');

$configurator->addConfig(__DIR__ . '/config/config.neon');
$configurator->addConfig(__DIR__ . '/config/config.local.neon');

$container = $configurator->createContainer();

return $container;
```

**Comment by dg** [comment]
I pushed an alternative and a slightly easier solution.

**Comment by dg** [comment]
@milo 

```php
	public static function boot(): Nette\Configurator
	{
		$configurator = new Nette\Configurator;

		# REST API
		if (strpos($_SERVER['HTTP_ACCEPT'] ?? '', 'application/json') !== false) {
			$configurator->setDebugMode(false);

		# CLI
		} elseif (PHP_SAPI === 'cli') {
			$configurator->setDebugMode(true);

		# Web
		} else {
			$configurator->setDebugMode([
				'192.168.0.2', '192.168.1.2', 
			]);
		}
		$configurator->enableTracy(__DIR__ . '/../log', 'miloslav.hula@...');

		$configurator->setTimeZone('Europe/Prague');
		$configurator->setTempDirectory(__DIR__ . '/../temp');

		$configurator->createRobotLoader()
			->addDirectory(__DIR__)
			->addDirectory(__DIR__ . '/../lib')
			->register();

		array_push(Tracy\Debugger::getBlueScreen()->keysToHide, 'clientSecret', 'secretKey', 'signatureKey');

		$configurator->addConfig(__DIR__ . '/config/config.neon');
		$configurator->addConfig(__DIR__ . '/config/config.local.neon');

		return $configurator; 
	}
```

**Comment by milo** [comment]
@dg I see 😄  Don't know why, I thought that the `Bootstrap` class will be part of `Nette\Application` namespace.

👍 from me

**Comment by dg** [comment]
(discussion on Slack https://pehapkari.slack.com/archives/C2R30BLKA/p1548981169298400 )

**Comment by dg** [comment]
Another suggestion is to rename config files to `common.neon` & `local.neon`

**Comment by fprochazka** [comment]
This might really shine if you make it easier to reuse the bootstrap between app and tests.

File: www/index.php [hunk 1]
```diff
@@ -2,7 +2,9 @@
 
 declare(strict_types=1);
 
-$container = require __DIR__ . '/../app/bootstrap.php';
+require __DIR__ . '/../vendor/autoload.php';
 
-$container->getByType(Nette\Application\Application::class)
+App\Application::boot()
+	->createContainer()
+	->getByType(Nette\Application\Application::class)
```
Comment (JanTvrdik) [inline www/index.php]: using two different classes named `Application` in once file si a terrible idea.

Comment (dg) [reply www/index.php]: Suggest better name.

Comment (enumag) [reply www/index.php]: Kernel?

**Comment by milo** [comment]
This is neat! Already switched to this new style few apps. `boot()`, `bootForCli()`, `bootForCron()`, `bootForTests()`... easy to manage.

**Comment by dg** [comment]
Change Booting back to Bootstrap?

Thread: https://twitter.com/JakubBoucek/status/1116084594711977986

**Comment by milo** [comment]
I'm OK with both. It's fact that `Bootstrap` is kind of convention.

**Comment by jakubboucek** [comment]
Thanks!

**Pull request merged**